marsha/core/api.py

```diff
@@ -41,8 +46,23 @@ def upload_confirm(request):
     if serializer.is_valid() is not True:
         return Response(serializer.errors, status=400)
 
-    # pylint: disable=fixme
-    # FIXME: check the validity of the signature once it is implemented in AWS
+    # The signed message is the s3 object key, any secret from the settings is valid
+    msg = serializer.validated_data["key"]
+    secrets = json.loads(settings.UPLOAD_CONFIRM_SHARED_SECRETS)
+
+    # Generate all possible valid signatures based on our list of secrets
+    # We need to do this to support 2 or more versions of our infrastructure at the same time.
+    # It then enables us to do updates and change the secret without incurring downtime.
+    valid_signatures = [
+        hmac.new(
+            secret.encode("utf-8"), msg=msg.encode("utf-8"), digestmod=hashlib.sha256
+        ).hexdigest()
+        for secret in secrets
+    ]
+
+    # Reject the request when the signature is not v
+    if serializer.validated_data["signature"] not in valid_signatures:
+        return Response("Forbidden", status=403)
```

Comment:
Since computing signatures takes time, it's more efficient to continue as soon as you have a positive. Something like:
```
signature_is_valid = any([
    serializer.validated_data["signature"] == hmac.new(
            secret.encode("utf-8"), msg=msg.encode("utf-8"), digestmod=hashlib.sha256
        ).hexdigest()
        for secret in secrets
])
```

It will evaluate to True as soon as there's a match and stop iterating.

Reply:
Nice! I did not know python provided such facilities in the stdlib